File: Sort/MergeSort.py
data = [10,12,18,1,3,5,9,26,32,41,7]

def merge(data1:list, data2:list):
    l = 0
    r = 0
    result = []
    while(l < len(data1) and r < len(data2)):
        if data1[l] <= data2[r]:
            result.append(data1[l])
            l+=1
        else:
            result.append(data2[r])
            r+=1
    if(l < len(data1)):
        result += data1[l:]
    if(r < len(data2)):
        result += data2[r:]
    return result
        

def MergeSort(data:list):
    if len(data) == 1:
        return data
    l = MergeSort(data[:int(len(data)/2)])
    r = MergeSort(data[int(len(data)/2) :])
    return merge(l,r)

print(MergeSort(data))

File: Sort/mergeSort.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <unordered_map>
using namespace std;

vector<int> mergeSort(vector<int> data, int l, int r) {
    if (l == r) {
        return vector<int>{data[l]};
    }

    int m = l + (r - l) / 2;
    vector<int> data1 = mergeSort(data, l, m);
    vector<int> data2 = mergeSort(data, m + 1, r);

    int l1 = data1.size();
    int l2 = data2.size();
    int i1 = 0, i2 = 0, k = 0;
    vector<int> result(l1 + l2);

    // Merge step
    while (i1 < l1 && i2 < l2) {
        if (data1[i1] <= data2[i2]) {
            result[k++] = data1[i1++];
        } else {
            result[k++] = data2[i2++];
        }
    }

    // Append remaining elements
    while (i1 < l1) {
        result[k++] = data1[i1++];
    }
    while (i2 < l2) {
        result[k++] = data2[i2++];
    }

    return result;
}

int main(){
    vector<int> arr = {5, 2, 9, 1, 6, 3};
    vector<int> sorted = mergeSort(arr, 0, arr.size() - 1);

    for (int num : sorted) {
        cout << num << " ";
    }

    return 0;
}